### PR TITLE
[Core] Retroactive signals.

### PR DESCRIPTION
**Pull request by Kobolog**
Use retroactive signals for service lifetime events in Context so that new subscribers to those signals are up-to-date with service lifetimes.

### CODE REVIEW

**Comment by Kobolog** [comment]
GCC 4.6 can't compile it and it doesn't have argument lifetime tracking yet, though.

**Comment by Kobolog** [comment]
Also, one problem with both implementations is that service.removed signal will not be called retroactively at all.

File: include/cocaine/signal.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+public:
+    typedef typename signal_type::slot_type slot_type;
+
+    auto
+    connect(const slot_type& slot) -> signals::connection {
+        {
+            std::lock_guard<std::mutex> guard(mutex);
+
+            std::for_each(history.begin(), history.end(), [&slot](const std::tuple<Args...>& args) {
+                slot(std::get<Indices>(args)...);
```
Comment (antmat) [inline include/cocaine/signal.hpp]: If I understand correctly invoking signal with connected expired slot is no-op, however calling expired slot explicitly throws. In this case here should be try-catch block.
Also there is a race condition here: 
call connect. stop after guard destruction
call operator() fully.
As a result connected slot will not be invoked with last argument pack.
This does not reallly affect current implementation but for safety I would lock entire connect and operator()

Comment (Kobolog) [inline include/cocaine/signal.hpp]: Good catch about the exception safety bug. As for the race, it is actually undefined behavior in underlying boost::signals implementation, so I left it out here as well.

**Pull request merged**